tempesta_fw/http_parser.c

```diff
@@ -604,11 +605,49 @@ __FSM_STATE(prefix ## _Done) {						\
 	return TFW_BLOCK;						\
 }
 
-/*
- * Read LWS at arbitrary position and move to stashed state.
- * This is bit complicated (however you can think about this as
- * a plain pushdown automaton), but reduces FSM code size.
+/* In request we should pass empty headers:
+ * RFC 7230 5.4:
+ * ....
+ * ....
+ * If the authority component is missing or
+ * undefined for the target URI, then a client MUST send a Host header
+ * field with an empty field-value.
+ *
+ * NOTE: using of RGEN_LWS_empty should be matched with
+ * the BUG_ON() statements in tfw_http_msg_hdr_val function
  */
+#define RGEN_LWS_common_cases(st)				\
+case '\n':							\
+	if (likely(!(parser->flags & TFW_HTTP_PF_LF))) {	\
+		parser->flags |= TFW_HTTP_PF_LF;		\
+		__FSM_MOVE(st);					\
+	}							\
+	return TFW_BLOCK;					\
+case ' ':							\
+case '\t':							\
+	__FSM_MOVE(st);						\
+default:							\
+	parser->flags &= ~TFW_HTTP_PF_CRLF;			\
+	parser->state = parser->_i_st;				\
+	parser->_i_st = 0;					\
+	BUG_ON(unlikely(p >= data + len || !*p));		\
+	goto fsm_reenter;
+
+#define RGEN_LWS_empty()						\
+__FSM_STATE(RGen_LWS_empty) {						\
+	switch (c) {							\
+	case '\r':							\
+		if (unlikely(parser->flags & TFW_HTTP_PF_LF))		\
+			return TFW_BLOCK;				\
+		parser->flags |= TFW_HTTP_PF_CR;			\
+		tfw_http_msg_hdr_chunk_fixup(msg, data, p - data);	\
+		if (tfw_http_msg_hdr_close(msg, parser->_hdr_tag))	\
+			return TFW_BLOCK;				\
+		__FSM_MOVE(RGen_LF);					\
+	RGEN_LWS_common_cases(RGen_LWS_empty)				\
+	}								\
+}
+
```

Comment:
The change should be bit more complex.

The initial ```RGEN_LWS()``` was written by RFC 2616 which allows line folding for headers. I.e. following request is correct in terms of RFC 2616:

        GET / HTTP/1.1\r\n
        Host:    \r\n
                foo.com\r\n
        Connection: close\r\n
        \r\n

Note that spaces precedes "foo.com" and this is required by RFC 2616. However, the logic is broken in initial ```RGEN_LWS()``` and "Connection: close" will be treated as Host header value in following request:

        GET / HTTP/1.1\r\n
        Host:\r\n
        Connection: close\r\n
        \r\n

Moreover, the feature is prohibited in RFC 7230:

        This specification deprecates such line folding except within the message/http
        media type (Section 8.3.1).

We don't use the parser to process body of ```message/http``` content type. RFC 7230 defines ```field-value``` as ```*( field-content / obs-fold )``` which could be zero length according to ABNF, RFC 5234, chapter 3.6. So we should process zero-length HTTP headers normally.